quinn-proto/src/connection/mod.rs

```diff
@@ -2747,6 +2737,40 @@ where
         sent
     }
 
+    /// Write pending ACKs into a buffer
+    ///
+    /// This method assumes ACKs are pending, and should only be called if
+    /// `PendingAcks::can_send()` or `!PendingAcks::ranges().is_empty()` returns `true`.
+    fn populate_acks(
```

Comment:
Maybe move this into `PacketSpace`? Methods without a receiver are a bit of an anti-pattern IMO, and we're always trying to move code out of `Connection` since it's already so large. I understand that means that `PacketSpace` now becomes aware of `SentFrames` and `ConnectionStats`, but that seems fine to me.

Reply:
Although the module is definitely on the too big side, I think `populate_acks` and `populate_packet` are currently extremely similar in style and touch the same data - therefore I don't think it's better to separate them. Also `spaces.rs` currently mostly contains datastructures, and less of serialization methods.

Reply:
Note that the non-receiver-ness here is just a workaround because it's called when a space is borrowed. It could be refactored to support a receiver by passing the space ID instead, and restarting the borrow afterwards in the caller, though in this instance the difference is minor.